ext/CUDASupportExt_FT.jl

```diff
@@ -0,0 +1,102 @@
+module CUDASupportExt_FT
+using CUDA 
+using Adapt
+using FourierTools
+using Base 
+# using NFFT
+# using CuNFFT
+
+# define a number of Union types to not repeat all definitions for each type
+AllShiftedType = Union{FourierTools.FourierSplit{<:Any,<:Any,<:Any},
+                            FourierTools.FourierJoin{<:Any,<:Any,<:Any}}
+
+# these are special only if a CuArray is wrapped
+
+AllSubArrayType = Union{SubArray{<:Any, <:Any, <:AllShiftedType, <:Any, <:Any},
+                        Base.ReshapedArray{<:Any, <:Any, <:AllShiftedType, <:Any},
+                        SubArray{<:Any, <:Any, <:Base.ReshapedArray{<:Any, <:Any, <:AllShiftedType, <:Any}, <:Any, <:Any}}
+AllShiftedAndViews = Union{AllShiftedType, AllSubArrayType}
+
+AllShiftedTypeCu{N, CD} = Union{FourierTools.FourierSplit{<:Any,<:Any,<:CuArray{<:Any,N,CD}},
+                            FourierTools.FourierJoin{<:Any,<:Any,<:CuArray{<:Any,N,CD}}}
+AllSubArrayTypeCu{N, CD} = Union{SubArray{<:Any, <:Any, <:AllShiftedTypeCu{N,CD}, <:Any, <:Any},
+                                 Base.ReshapedArray{<:Any, <:Any, <:AllShiftedTypeCu{N,CD}, <:Any},
+                                 SubArray{<:Any, <:Any, <:Base.ReshapedArray{<:Any, <:Any, <:AllShiftedTypeCu{N,CD}, <:Any}, <:Any, <:Any}}
+AllShiftedAndViewsCu{N, CD} = Union{AllShiftedTypeCu{N, CD}, AllSubArrayTypeCu{N, CD}}
+
+Adapt.adapt_structure(to, x::FourierTools.FourierSplit{T, M, AA, D}) where {T, M, AA, D} = FourierTools.FourierSplit(adapt(to, parent(x)), Val(D), x.L1, x.L2, x.do_split);
+Adapt.adapt_structure(to, x::FourierTools.FourierJoin{T, M, AA, D}) where {T, M, AA, D} = FourierTools.FourierJoin(adapt(to, parent(x)), Val(D), x.L1, x.L2, x.do_join);
+
+function Base.Broadcast.BroadcastStyle(::Type{T})  where {N, CD, T<:AllShiftedTypeCu{N, CD}}
+    CUDA.CuArrayStyle{N,CD}()
+end
+
+# Define the BroadcastStyle for SubArray of MutableShiftedArray with CuArray
+
+function Base.Broadcast.BroadcastStyle(::Type{T})  where {N, CD, T<:AllSubArrayTypeCu{N, CD}}
+    CUDA.CuArrayStyle{N,CD}()
+end
+
+function Base.copy(s::AllShiftedAndViews)
```

Comment:
this is type piracy?

Because you define a copy method from Base with types owned by base. So this should not be done.

Reply:
I don't think this is type piracy here. `AllShiftedAndViews` only contains types or unions which are defined in this package.  However there are arguments one could make with the `CUDASupportExt_SA.jl`. In that case, I really don't know what to do, since the author of `ShiftedArrays.jl` is non-responsive. 
https://github.com/JuliaArrays/ShiftedArrays.jl/pull/70
https://github.com/JuliaArrays/ShiftedArrays.jl/issues/64
Maybe I just dublicate the code for the `circshiftedArray.jl` into `MutableShiftedArrays.jl` and host the extension there and then remove the dependence on ShiftedArrays.jl entirely.?